Handle dot in project name correctly without dropping part of the project name

fletcher-sumglobal opened the pull request:
Issue [#960](https://github.com/gradle/gradle-native/issues/960)
Handle dot in project name without dropping part of the project name.

Signed-off-by: Kent Fletcher <kfletcher@sumglobal.com>

### Context
<!--- Why do you believe many users will benefit from this change? -->
<!--- Link to relevant issues or forum discussions here -->

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

fletcher-sumglobal commented:
@lacasseio Thanks for pointing that out.  I'll get right on it.

fletcher-sumglobal commented:
I made another commit refactoring of the code to add overloaded methods for the getters of the interface `PlatformToolProvider` -> and implemented in `AbtractPlatformToolProvider` & `UnavailablePlatformToolProvider`.  Also added unit tests for these getters.  Now I'm having a problem with integration tests.  It is not so apparent how to test these overloaded methods in the spock framework.  I'm working on trying to figure that out.

fletcher-sumglobal commented:
@lacasseio In hindsight, after closely examining the integration  test, I've found the PlatformToolProvider 'getters' are mocked in the exact same way it is coded in the real or actual implementation of my first [commit](https://github.com/gradle/gradle/pull/7959/commits/5f153d1959d8154f1be7577b14eb4d48e6f3681c).  In other words, it is mocked as 'some/path/to/file/baseName' + '.ext', just a concatenation of the two strings, and not using 'withExtension' method at all.  The only code that rely's on the 'withExtension' is the unit test for OperatingSystem which are really superfluous assertions, and not like what is happening in reality in the code base for a real project.  No native code projects to my knowledge depends on the removal of an extension, then adding it back to the returned string.  That said, I've added the overloaded methods anyway, 'just in case' there are some projects that do rely on it.  Plus I've added unit tests to the OperatingSystemTest accordingly.  They are in my humble opinion, some of the assertions are superfluous. 

However, now that I've gone down this route,  I'm having a problem updating the NativeBasePluginTest integration test.  I've updated the mock closure expressions (according the the spock documentation) for platformToo.getXxxx() to take two args like the real code does, but during a check of integ test execution, the mocked method returns an empty string  "" causing the unit test to fail.  I can't for the life of me figure out why it is failing to return the concatenated string.  I was wondering if you could take a look at the following code to see what is wrong with the mocked method closures for platformTool.getXxx().  Again, it looks right according to spock documentation, but the closure doesn't return my expected results, in fact, I don't think spock even calls the mocked closure.  Here are the lines of code here -> [NativeBasePluginTest:line-156](https://github.com/gradle/gradle/blob/78a278c55b3f996b40884499dd96a962ffcd114e/subprojects/language-native/src/test/groovy/org/gradle/language/plugins/NativeBasePluginTest.groovy#L156)), [NativeBasePluginTest:line-192](https://github.com/gradle/gradle/blob/78a278c55b3f996b40884499dd96a962ffcd114e/subprojects/language-native/src/test/groovy/org/gradle/language/plugins/NativeBasePluginTest.groovy#L192), [NativeBasePluginTest:line-228](https://github.com/gradle/gradle/blob/78a278c55b3f996b40884499dd96a962ffcd114e/subprojects/language-native/src/test/groovy/org/gradle/language/plugins/NativeBasePluginTest.groovy#L228), [NativeBasePluginTest:line-231](https://github.com/gradle/gradle/blob/78a278c55b3f996b40884499dd96a962ffcd114e/subprojects/language-native/src/test/groovy/org/gradle/language/plugins/NativeBasePluginTest.groovy#L231), [NativeBasePluginTest:line-280](https://github.com/gradle/gradle/blob/78a278c55b3f996b40884499dd96a962ffcd114e/subprojects/language-native/src/test/groovy/org/gradle/language/plugins/NativeBasePluginTest.groovy#L280), [NativeBasePluginTest:line-333](https://github.com/gradle/gradle/blob/78a278c55b3f996b40884499dd96a962ffcd114e/subprojects/language-native/src/test/groovy/org/gradle/language/plugins/NativeBasePluginTest.groovy#L333),[NativeBasePluginTest:line-336](https://github.com/gradle/gradle/blob/78a278c55b3f996b40884499dd96a962ffcd114e/subprojects/language-native/src/test/groovy/org/gradle/language/plugins/NativeBasePluginTest.groovy#L336)

There may be a  bug in spock where it can't handle mocking up methods that are overloaded.

fletcher-sumglobal commented:
I think I found the answer to my question...  In order to for spock to mock the method with a closure to support mutliple arguments to my overloaded method, I need to add `*_` for my argument constraints.  Previously, in my mock method, I had:
```
        toolProvider.getStaticLibraryName(_) >> { String d, String baseName ->
            d + baseName + ".lib"
        }
```
Instead of `'_'`, I needed to use `'*_'` as follows:
```
        toolProvider.getStaticLibraryName(*_) >> { String d, String baseName ->
            d + baseName + ".lib"
        }
```

Per docs
```
1 * subscriber.receive(_)           // any single argument (including null)
1 * subscriber.receive(*_)          // any argument list (including the empty argument list)
```

fletcher-sumglobal commented:
@lacasseio all unit tests, and integration tests are complete to my knowledge.   I don't know what you guys think about this, but maybe instead of overloading the method, just pick one method or the other?  I don't really know of a need for having both ways of doing things because I really don't think the 'current' behavior was ever desirable in the first place.  The only place I could find that depended on the current behavior was unit tests.  And the tests that are currently written, the assertions don't reflect a real life use case.  But then again, I don't have near the level of experience you guys have in all the native gradle code.

lacasseio commented:
I think this PR went in the wrong direction. We shouldn't have to change the test for this PR to go through. The concern I had was if we change the implementation in one place, we need to be mindful of every other place the code is used since it will cause failures. What I see that is missing is integration tests showing project named with a dot in them. We don't want to change the current tests to provide coverage for this work as we would be removing coverage for other features. Since `OperatingSystem` is internal, we can change the implementation. We don't want to have a duplicated API just in case.

Instead, we want to have an acceptable implementation that covers both possible outcomes. Something to look into is if we change the implementation, what tests are failing. According to those failures, is the test miss written or are we changing a public behavior? Given the public behavior, is it an acceptable change that we can push in the next version of Gradle as a possible breaking change?  For example, we want to do a basic check to see if the extension is already present on the path to avoid duplication which is what original implementation was doing. It may be acceptable that we append the extension over another extension to avoid causing the issue we are trying to fix. What we need to see is what is the impact across all native. It looks like an acceptable change, but for this, we would need to do the change.

The key here is to always start with a test of some sort. In this situation, an integration test that targets C++ application and library, as well as Swift application and library, would be the right start. In those test, we assert the path match what is expected (full project name with dots is contained within). Then the implementation can proceed with the care of not breaking other tests. If the other tests break, we need to evaluate if it makes sense to change the test or change the implementation. The majority of the time the test should not be changed because they express something the user expects. However, we do have tests that aren't asserting the right things and those can safely be changed.

fletcher-sumglobal commented:
Changing the existing integration test was necessary because the underlying implementation changed for the PlatformToolProvder getters, and their usage in NativeBasePlugin.  Without updating the mocked impl, the integration test effectively fails because the mocked method closure no longer matches the signature of the method invoked inside NativeBasePlugin.  So changing this part of the integ test was necessary.

But in reference to stripping off the extension, then appending it back, this very same closure, does NOT do this.  The closure  `toolProvider.getStaticLibraryName(_) >> {String s -> s + ".lib" }` overrides the internal behavior of `OperatingSystem`.  As you can clearly see, this closure doesn't check for extension before appending, it is merely concatenating strings.   Effectively, this integration test masks the true underlying runtime behavior with this closure, but it also happens to be the expected result.  So if anything, this PR is aligning the implementation to more closely match the integration test.  So I think you are correct in assuming what is in the integration test is what the user expects, and based on close examination of these test indicate that the current behavior isn't what is expected here.   So I don't see any change in public behavior here unless someone has intentionally set their base project name in `settings.gradle` to `foo.dll`, then these new changes would effectively create build output with `foo.dll.dll`.

Other than adding additional variations of the integration test per: 
> plus an integration test that targets C++ application and library, as well as Swift application and library

I'm not entirely clear on what direction you want me to take here.  Based on Sterling's comments, I thought overloading the methods might be the best approach given what I knew at the time.  But, to be perfectly honest, I think my first commit in this PR is the best solution, and that was simply removing the 'withExtension' method call.  I don't see any code that relies on it except some superfluous unit tests in baseServices.  What do you think about the method overloading vs removal of 'withExtension' altogether.  I don't see a way to accommodate both behaviors per your comment:
> we want to have an acceptable implementation that covers both possible outcomes

fletcher-sumglobal commented:
Overloading these getters is turning out way more painful than it needs to be.  This requires modifying all the test fixtures to break apart the 'parentDir' & 'baseName' as two parameters as opposed to passing the full project path in as it is currently done.  Rather than go through all this effort, I would rather go with my first commit  in this as I see it having the least risk.

fletcher-sumglobal commented:
I can write integration tests much easier (with 'dot' in project name for cpp-application/cpp-library) with using only the first commit 5f153d1 in this PR.  I know that this code works apart from affecting a couple assertions in the unit test  The overloading is getting too painful, and I don't think you guys are keen on that idea to begin with.

I'm going to rework this PR with a squash commit to a variation of the first commit, and show some integration tests along with it for your review.   This is getting too ugly.

fletcher-sumglobal commented:
@lacasseio The most recent commit ->  337268e is a squash commit force push, and supercedes all previous commits..   

This PR started going sideways on me, and got out of hand overloading those methods in PlatformToolProvider.  Its impact was far greater than I expected.  This is a squash commit simplifying the PR down to a variant of my first commit, while still retaining the functionality 'withExtension' stripping/appending extensions off when needed.  It appears a number of integration test depended on this extension swapping behavior, so to retain it, I implemented a better version of 'withExtension' within OperatingSystem and slapped a regex condition on the filepath to continue to allow both behaviors, retaining the extension swap for cases that depended on it, and ignoring the swap if the filepath didn't meet the regex criteria.

Also, made sure all existing unit/integration test pass, while also adding new ones as requested in the PR.  Although I added a swift integ test, I couldn't execute it locally, so maybe the CI can run it to see results.
